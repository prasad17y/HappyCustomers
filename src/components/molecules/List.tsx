import React from 'react';
import {
  SectionList,
  StyleSheet,
  RefreshControl,
  SectionListData,
} from 'react-native';

interface ListProps<ItemT, SectionT> {
  sections: ReadonlyArray<SectionListData<ItemT, SectionT>>;
  renderItem: ({item}: {item: ItemT}) => React.ReactElement | null;
  renderSectionHeader: (info: {
    section: SectionListData<ItemT, SectionT>;
  }) => React.ReactElement | null;
  refreshing: boolean;
  onRefresh: () => void;
  keyExtractor: (item: ItemT, index: number) => string;
  ListFooterComponent?: React.ComponentType<any> | React.ReactElement | null;
  ListEmptyComponent?: React.ComponentType<any> | React.ReactElement | null;
}

const ITEM_HEIGHT = 64;
const SECTION_HEADER_HEIGHT = 34;

const List = <ItemT, SectionT>({
  sections,
  renderItem,
  renderSectionHeader,
  refreshing,
  onRefresh,
  keyExtractor,
  ListFooterComponent,
  ListEmptyComponent,
}: ListProps<ItemT, SectionT>) => {
  const getItemLayout = React.useCallback(
    (
      data: SectionListData<ItemT, SectionT>[] | null,
      index: number,
    ): {length: number; offset: number; index: number} => {
      if (!data) {
        return {length: 0, offset: 0, index};
      }

      let offset = 0;
      let absoluteIndex = 0;

      for (const section of data) {
        if (index === absoluteIndex) {
          return {length: SECTION_HEADER_HEIGHT, offset, index};
        }
        offset += SECTION_HEADER_HEIGHT;
        absoluteIndex += 1;

        if (index < absoluteIndex + section.data.length) {
          const itemIndexInSection = index - absoluteIndex;
          return {
            length: ITEM_HEIGHT,
            offset: offset + itemIndexInSection * ITEM_HEIGHT,
            index,
          };
        }

        offset += section.data.length * ITEM_HEIGHT;
        absoluteIndex += section.data.length;
      }

      return {length: 0, offset, index};
    },
    [],
  );

  return (
    <SectionList
      sections={sections}
      keyExtractor={keyExtractor}
      renderItem={renderItem}
      renderSectionHeader={renderSectionHeader}
      contentContainerStyle={
        sections.length > 0 ? styles.listContent : styles.emptyContent
      }
      stickySectionHeadersEnabled={false}
      initialNumToRender={15}
      maxToRenderPerBatch={10}
      windowSize={5}
      getItemLayout={getItemLayout}
      refreshControl={
        <RefreshControl refreshing={refreshing} onRefresh={onRefresh} />
      }
      ListEmptyComponent={ListEmptyComponent}
      ListFooterComponent={ListFooterComponent}
    />
  );
};

const styles = StyleSheet.create({
  listContent: {
    paddingBottom: 20,
  },
  emptyContent: {
    flex: 1,
  },
});

export default React.memo(List) as typeof List;
