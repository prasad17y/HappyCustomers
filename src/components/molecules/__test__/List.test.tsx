import React from 'react';
import {render, screen, fireEvent} from '@testing-library/react-native';
import List from '../List';
import {Text, View} from 'react-native';

const mockSections = [
  {
    title: 'A',
    data: [
      {id: '1', name: 'Alice'},
      {id: '2', name: 'Alicia'},
    ],
  },
  {
    title: 'B',
    data: [{id: '3', name: 'Bob'}],
  },
];

// Mock render components
const renderItem = ({item}: {item: {id: string; name: string}}) => (
  <View>
    <Text>{item.name}</Text>
  </View>
);

const renderSectionHeader = ({section}: {section: any}) => (
  <View>
    <Text>{section.title}</Text>
  </View>
);

const keyExtractor = (item: {id: string}) => item.id;

describe('List', () => {
  it('should render sections and items correctly', () => {
    render(
      <List
        sections={mockSections}
        renderItem={renderItem}
        renderSectionHeader={renderSectionHeader}
        refreshing={false}
        onRefresh={jest.fn()}
        keyExtractor={keyExtractor}
      />,
    );

    // Check for section headers
    expect(screen.getByText('A')).toBeTruthy();
    expect(screen.getByText('B')).toBeTruthy();

    // Check for list items
    expect(screen.getByText('Alice')).toBeTruthy();
    expect(screen.getByText('Alicia')).toBeTruthy();
    expect(screen.getByText('Bob')).toBeTruthy();
  });

  it('should render the ListEmptyComponent when sections are empty', () => {
    const EmptyComponent = () => <Text>The list is empty.</Text>;
    render(
      <List
        sections={[]}
        renderItem={renderItem}
        renderSectionHeader={renderSectionHeader}
        refreshing={false}
        onRefresh={jest.fn()}
        keyExtractor={keyExtractor}
        ListEmptyComponent={EmptyComponent}
      />,
    );

    expect(screen.getByText('The list is empty.')).toBeTruthy();
  });

  it('should render the ListFooterComponent', () => {
    const FooterComponent = () => <Text>End of list.</Text>;
    render(
      <List
        sections={mockSections}
        renderItem={renderItem}
        renderSectionHeader={renderSectionHeader}
        refreshing={false}
        onRefresh={jest.fn()}
        keyExtractor={keyExtractor}
        ListFooterComponent={FooterComponent}
      />,
    );

    expect(screen.getByText('End of list.')).toBeTruthy();
  });

  it('should call onRefresh when a pull-to-refresh is simulated', () => {
    const onRefreshMock = jest.fn();
    render(
      <List
        sections={mockSections}
        renderItem={renderItem}
        renderSectionHeader={renderSectionHeader}
        refreshing={false}
        onRefresh={onRefreshMock}
        keyExtractor={keyExtractor}
      />,
    );

    fireEvent(screen.getByTestId('sectionlist'), 'refresh');

    expect(onRefreshMock).toHaveBeenCalledTimes(1);
  });

  it('should match the snapshot', () => {
    const tree = render(
      <List
        sections={mockSections}
        renderItem={renderItem}
        renderSectionHeader={renderSectionHeader}
        refreshing={false}
        onRefresh={jest.fn()}
        keyExtractor={keyExtractor}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});

describe('List getItemLayout', () => {
  const ITEM_HEIGHT = 64;
  const SECTION_HEADER_HEIGHT = 34;

  const {getByTestId} = render(
    <List
      sections={mockSections}
      renderItem={renderItem}
      renderSectionHeader={renderSectionHeader}
      refreshing={false}
      onRefresh={jest.fn()}
      keyExtractor={keyExtractor}
    />,
  );

  // The underlying SectionList is given a testID by default in RNTL, but we can find it by role
  const sectionListInstance = getByTestId('sectionlist');
  const getItemLayout = sectionListInstance.props.getItemLayout;

  it('should calculate the layout for the first section header (index 0)', () => {
    const layout = getItemLayout(mockSections, 0);
    expect(layout.offset).toBe(0);
    expect(layout.length).toBe(SECTION_HEADER_HEIGHT);
  });

  it('should calculate the layout for the first item in the first section (index 1)', () => {
    const layout = getItemLayout(mockSections, 1);
    expect(layout.offset).toBe(SECTION_HEADER_HEIGHT);
    expect(layout.length).toBe(ITEM_HEIGHT);
  });

  it('should calculate the layout for the second item in the first section (index 2)', () => {
    const layout = getItemLayout(mockSections, 2);
    const expectedOffset = SECTION_HEADER_HEIGHT + ITEM_HEIGHT; // 34 + 64 = 98
    expect(layout.offset).toBe(expectedOffset);
    expect(layout.length).toBe(ITEM_HEIGHT);
  });

  it('should calculate the layout for the second section header (index 3)', () => {
    const layout = getItemLayout(mockSections, 3);
    const expectedOffset =
      SECTION_HEADER_HEIGHT + mockSections[0].data.length * ITEM_HEIGHT; // 34 + 2 * 64 = 162
    expect(layout.offset).toBe(expectedOffset);
    expect(layout.length).toBe(SECTION_HEADER_HEIGHT);
  });

  it('should calculate the layout for the first item in the second section (index 4)', () => {
    const layout = getItemLayout(mockSections, 4);
    const expectedOffset =
      SECTION_HEADER_HEIGHT + // Header A
      mockSections[0].data.length * ITEM_HEIGHT + // Items in A
      SECTION_HEADER_HEIGHT; // Header B
    // 34 + 2 * 64 + 34 = 196
    expect(layout.offset).toBe(expectedOffset);
    expect(layout.length).toBe(ITEM_HEIGHT);
  });
});
