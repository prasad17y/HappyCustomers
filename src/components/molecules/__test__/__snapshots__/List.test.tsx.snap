// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should match the snapshot 1`] = `
<RCTScrollView
  contentContainerStyle={
    {
      "paddingBottom": 20,
    }
  }
  data={
    [
      {
        "data": [
          {
            "id": "1",
            "name": "Alice",
          },
          {
            "id": "2",
            "name": "Alicia",
          },
        ],
        "title": "A",
      },
      {
        "data": [
          {
            "id": "3",
            "name": "Bob",
          },
        ],
        "title": "B",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  getItemLayout={[Function]}
  initialNumToRender={15}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[MockFunction]}
      refreshing={false}
    />
  }
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={[]}
  testID="sectionlist"
  windowSize={5}
>
  <RCTRefreshControl />
  <View>
    <View
      onFocusCapture={[Function]}
      style={null}
    >
      <View>
        <Text>
          A
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      style={null}
    >
      <View>
        <Text>
          Alice
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      style={null}
    >
      <View>
        <Text>
          Alicia
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      style={null}
    />
    <View
      onFocusCapture={[Function]}
      style={null}
    >
      <View>
        <Text>
          B
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      style={null}
    >
      <View>
        <Text>
          Bob
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      style={null}
    />
  </View>
</RCTScrollView>
`;
